drivers: rtc: rpi_pico: fix tm_mon and tm_mday offsets

kartben opened the pull request:
tm_mon is 0-11 and pico hw is 1-12
tm_mday is 1-31 and so is pico

Updated RTC tests so that they are actually testing rollover scenarios for alarms. @bjarki-andreasen please review -- I am guessing this will make quite a few drivers unhappy as I doubt all of them just happen to have perfectly addressed these if there wasn't coverage for it :) FWIW the test still passes for `zephyr,rtc-emul` so at least there's that!

kartben committed:
drivers: rtc: rpi_pico: fix tm_mon and tm_mday offsets

tm_mon is 0-11 and pico hw is 1-12
tm_mday is 1-31 and so is pico

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

ajf58 commented:
Is there a failing unit test that goes with this? It's been a while since I looked at this, but wouldn't that also mean that the logic rtc_rpi_pico_get_time is wrong?

kartben commented:
> Is there a failing unit test that goes with this? It's been a while since I looked at this, but wouldn't that also mean that the logic rtc_rpi_pico_get_time is wrong?

Interestingly the code for get_time was actually correct https://github.com/zephyrproject-rtos/zephyr/blob/3172f85b08a07e47bb8ad87f2662a5b45080869d/drivers/rtc/rtc_rpi_pico.c#L146-L147 and so was set_time https://github.com/zephyrproject-rtos/zephyr/blob/3172f85b08a07e47bb8ad87f2662a5b45080869d/drivers/rtc/rtc_rpi_pico.c#L116-L117

kartben committed:
drivers: tests: rtc: harden alarm tests

Ensure the alarm tests exercise rollover scenarios by setting RTC time
to a few seconds before midnight on a December 31st and alarm time to a
few seconds after midnight on the following day.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
drivers: tests: rtc: add overlay for adafruit_macropad_rp2040

Allow to run the RTC test suite on Adafruit Macropad RP2040

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben commented:
@ajf58 I will be updating the RTC unit test so that it showcases the issue -- right now it does not hit any "roll over" scenario so the bug wouldn't be caught.

kartben committed:
drivers: tests: rtc: relax y2k test seconds check

When testing y2k rollover, the seconds value can easily be off by one
since there is no guarantee that the RTC will be exactly on the second
boundary.
Relax the check to allow for a one second difference.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben commented:
```
west build -b adafruit_macropad_rp2040 tests/drivers/rtc/rtc_api -p
```

With the last commit in this PR:

```
Running TESTSUITE rtc_api
===================================================================
I: Setting alarm
 PASS - test_alarm in 26.005 seconds
===================================================================
START - test_alarm_callback
I: Setting alarm
I: Setting alarm
 PASS - test_alarm_callback in 26.003 seconds
===================================================================
START - test_set_get_time
 PASS - test_set_get_time in 0.001 seconds
===================================================================
START - test_time_counting
 PASS - test_time_counting in 9.003 seconds
===================================================================
START - test_y2k
 PASS - test_y2k in 2.001 seconds
===================================================================
TESTSUITE rtc_api succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [rtc_api]: pass = 5, fail = 0, skip = 0, total = 5 duration = 63.013 seconds
 - PASS - [rtc_api.test_alarm] duration = 26.005 seconds
 - PASS - [rtc_api.test_alarm_callback] duration = 26.003 seconds
 - PASS - [rtc_api.test_set_get_time] duration = 0.001 seconds
 - PASS - [rtc_api.test_time_counting] duration = 9.003 seconds
 - PASS - [rtc_api.test_y2k] duration = 2.001 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

Without it:

```
Running TESTSUITE rtc_api
===================================================================
START - test_alarm
I: Setting alarm
I: Setting alarm

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/rtc/rtc_api/src/test_alarm.c:203: rtc_api_test_alarm: (ret not equal to 1)
Alarm 0 should be pending
 FAIL - test_alarm in 13.016 seconds
===================================================================
START - test_alarm_callback
I: Setting alarm

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/rtc/rtc_api/src/test_alarm_callback.c:140: rtc_api_test_alarm_callback: (callback_called_status not equal to true)
Alarm 0 callback should have been called
 FAIL - test_alarm_callback in 13.019 seconds
===================================================================
START - test_set_get_time
 PASS - test_set_get_time in 0.001 seconds
===================================================================
START - test_time_counting
 PASS - test_time_counting in 9.003 seconds
===================================================================
START - test_y2k
 PASS - test_y2k in 2.001 seconds
===================================================================
TESTSUITE rtc_api failed.

------ TESTSUITE SUMMARY START ------

SUITE FAIL -  60.00% [rtc_api]: pass = 3, fail = 2, skip = 0, total = 5 duration = 37.040 seconds
 - FAIL - [rtc_api.test_alarm] duration = 13.016 seconds
 - FAIL - [rtc_api.test_alarm_callback] duration = 13.019 seconds
 - PASS - [rtc_api.test_set_get_time] duration = 0.001 seconds
 - PASS - [rtc_api.test_time_counting] duration = 9.003 seconds
 - PASS - [rtc_api.test_y2k] duration = 2.001 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION FAILED
```

bjarki-andreasen reviewed (approved):
Nice addition :) We will probably get some more fix PRs once vendors run the tests downstream :)

kartben commented:
@ThreeEights thanks for your +1 :) ... which should be counting towards approvals. Please check you applied to be a Contributor as you don't seem to be properly attached to the zephyrproject-rtos GIthub org. Thanks!
